### PR TITLE
Allow multiple tuneable mutational stages

### PR DESCRIPTION
**Pull request by wtdcode**
I can't see obvious reasons to make `TuneableMutationalStage` a singleton class and it's super inconvenient that every fuzzer can only have one `TuneableMutationalStage`. Therefore, like the observers and feedbacks, I assign a name for each `TuneableMutationalStage` to allow multiple mutational strages while keeping maximum backward compatibility.

### CODE REVIEW

**Comment by domenukk** [comment]
This needs a `cargo fmt` I think, else it looks reasonable :) thanks

**Comment by wtdcode** [comment]
> This needs a `cargo fmt` I think, else it looks reasonable :) thanks

I already did that and I just tred `cargo fmt` again while nothing changed.

**Comment by domenukk** [comment]
```rust
Diff in /Users/runner/work/LibAFL/LibAFL/libafl/src/stages/tuneable.rs at line 1:
 //! A [`crate::stages::MutationalStage`] where the mutator iteration can be tuned at runtime
 
+use alloc::string::{String, ToString};
 use core::{marker::PhantomData, time::Duration};
 
-use alloc::string::{String, ToString};
 use libafl_bolts::{current_time, impl_serdeany, rands::Rand};
 use serde::{Deserialize, Serialize};
```

You may need `cargo +nighlty fmt` to reorder inputs

**Comment by wtdcode** [comment]
Turns out that I forgot to switch the branch

File: libafl/src/stages/tuneable.rs [hunk 1]
```diff
@@ -227,43 +239,50 @@ where
...
-    pub fn set_iters<S: HasMetadata>(state: &mut S, iters: u64) -> Result<(), Error> {
-        set_iters(state, iters)
+    pub fn set_iters<S: HasNamedMetadata>(&self, state: &mut S, iters: u64) -> Result<(), Error> {
+        set_iters(state, iters, &self.name)
     }
 
     /// Get the set iterations
-    pub fn iters<S: HasMetadata>(state: &S) -> Result<Option<u64>, Error> {
-        get_iters(state)
+    pub fn iters<S: HasNamedMetadata>(&self, state: &S) -> Result<Option<u64>, Error> {
```
Comment (domenukk) [inline libafl/src/stages/tuneable.rs:267]: Maybe have a `iters_named` fn then?

**Comment by domenukk** [comment]
@andreafioraldi is there overhead for named vs non named? Assuming this is on the hot path, would it make sense to have extra functions for non-named?

File: libafl/src/stages/tuneable.rs [hunk 2]
```diff
@@ -28,51 +29,61 @@ struct TuneableMutationalStageMetadata {
...
-        .get::<TuneableMutationalStageMetadata>()
+        .named_metadata_map()
+        .get::<TuneableMutationalStageMetadata>(name)
         .ok_or_else(|| Error::illegal_state("TuneableMutationalStage not in use"))
         .map(|metadata| metadata.fuzz_time)
 }
 
 /// Reset this to a normal, randomized, stage
-pub fn reset<S: HasMetadata>(state: &mut S) -> Result<(), Error> {
+pub fn reset<S: HasNamedMetadata>(state: &mut S, name: &str) -> Result<(), Error> {
```
Comment (andreafioraldi) [inline libafl/src/stages/tuneable.rs:83]: Can you move the trait bounds in the where clause? That's the style for most of the code in libafl

File: libafl/src/stages/tuneable.rs (hunk 1 above)
Comment (andreafioraldi) [reply libafl/src/stages/tuneable.rs:267]: IMO using named is just fine, this is called once per scheduled input IIRC, so not really an hot path.

**Review by andreafioraldi — CHANGES_REQUESTED** [review]

File: libafl/src/stages/tuneable.rs (hunk 2 above)
Comment (wtdcode) [reply libafl/src/stages/tuneable.rs:83]: Sure, I'm just following the old code

**Comment by domenukk** [comment]
I still don't like that now users _have_ to use a name, I might prefer a second API of `_named` for usability sake

**Comment by wtdcode** [comment]
Done

File: libafl/src/stages/tuneable.rs [hunk 3]
```diff
@@ -231,43 +295,55 @@ where
     }
 }
 
-impl<E, EM, M, Z> TuneableMutationalStage<E, EM, Z::Input, M, Z>
+impl<E, EM, I, M, Z> TuneableMutationalStage<E, EM, I, M, Z>
```
Comment (domenukk) [inline libafl/src/stages/tuneable.rs:298]: What's the reason to have an extra `I` here? Do you use it with different inputs?

**Comment by domenukk** [comment]
Sweet, thx :)

File: libafl/src/stages/tuneable.rs (hunk 3 above)
Comment (wtdcode) [reply libafl/src/stages/tuneable.rs:298]: Without the bound change, I can't call `transforming` due to different bounds.

**Pull request merged**